README.md
[![ci](https://github.com/sinlov/drone-feishu-group-robot/workflows/ci/badge.svg?branch=main)](https://github.com/sinlov/drone-feishu-group-robot/actions/workflows/ci.yml)

[![go mod version](https://img.shields.io/github/go-mod/go-version/sinlov/drone-feishu-group-robot?label=go.mod)](https://github.com/sinlov/drone-feishu-group-robot)
[![GoDoc](https://godoc.org/github.com/sinlov/drone-feishu-group-robot?status.png)](https://godoc.org/github.com/sinlov/drone-feishu-group-robot)
[![goreportcard](https://goreportcard.com/badge/github.com/sinlov/drone-feishu-group-robot)](https://goreportcard.com/report/github.com/sinlov/drone-feishu-group-robot)

[![docker hub version semver](https://img.shields.io/docker/v/sinlov/drone-feishu-group-robot?sort=semver)](https://hub.docker.com/r/sinlov/drone-feishu-group-robot/tags?page=1&ordering=last_updated)
[![docker hub image size](https://img.shields.io/docker/image-size/sinlov/drone-feishu-group-robot)](https://hub.docker.com/r/sinlov/drone-feishu-group-robot)
[![docker hub image pulls](https://img.shields.io/docker/pulls/sinlov/drone-feishu-group-robot)](https://hub.docker.com/r/sinlov/drone-feishu-group-robot/tags?page=1&ordering=last_updated)

[![GitHub license](https://img.shields.io/github/license/sinlov/drone-feishu-group-robot)](https://github.com/sinlov/drone-feishu-group-robot)
[![codecov](https://codecov.io/gh/sinlov/drone-feishu-group-robot/branch/FE-add-build-trigger-show/graph/badge.svg)](https://codecov.io/gh/sinlov/drone-feishu-group-robot)
[![GitHub latest SemVer tag)](https://img.shields.io/github/v/tag/sinlov/drone-feishu-group-robot)](https://github.com/sinlov/drone-feishu-group-robot/tags)
[![GitHub release)](https://img.shields.io/github/v/release/sinlov/drone-feishu-group-robot)](https://github.com/sinlov/drone-feishu-group-robot/releases)

## for what

- this project used to drone CI use [Custom bot guide](https://open.feishu.cn/document/ukTMukTMukTM/ucTM5YjL3ETO24yNxkjN?lang=en-US)
- support drone-docker `platform=linux/arm,linux/arm64,linux/amd64`
- support drone-exec

## Contributing

[![Contributor Covenant](https://img.shields.io/badge/contributor%20covenant-v1.4-ff69b4.svg)](.github/CONTRIBUTING_DOC/CODE_OF_CONDUCT.md)
[![GitHub contributors](https://img.shields.io/github/contributors/sinlov/drone-feishu-group-robot)](https://github.com/sinlov/drone-feishu-group-robot/graphs/contributors)

We welcome community contributions to this project.

Please read [Contributor Guide](.github/CONTRIBUTING_DOC/CONTRIBUTING.md) for more information on how to get started.

请阅读有关 [贡献者指南](.github/CONTRIBUTING_DOC/zh-CN/CONTRIBUTING.md) 以获取更多如何入门的信息

## Features

### v1.16.+

- show build time and Stage details info

![img.png](https://github.com/sinlov/drone-feishu-group-robot/raw/main/features/release-v1.16.x/pr-success.png)

### v1.15.+

- let feishu message card render HeadTemplateStyle support different event

- build status success

![](https://github.com/sinlov/drone-feishu-group-robot/raw/main/features/release-v1.15.x/push-success.png)

- build status failure

![](https://github.com/sinlov/drone-feishu-group-robot/raw/main/features/release-v1.15.x/push-failure.png)

- pull request status success

![](https://github.com/sinlov/drone-feishu-group-robot/raw/main/features/release-v1.15.x/pr-success.png)

- pull request status failure

![](https://github.com/sinlov/drone-feishu-group-robot/raw/main/features/release-v1.15.x/pr-failure.png)

- tag status success

![](https://github.com/sinlov/drone-feishu-group-robot/raw/main/features/release-v1.15.x/tag-success.png)

### more features

- see [features/README.md](https://github.com/sinlov/drone-feishu-group-robot/blob/main/features/README.md)

## before use

- sed doc at [feishu Custom bot guide](https://open.feishu.cn/document/ukTMukTMukTM/ucTM5YjL3ETO24yNxkjN?lang=en-US), to new group robot
- Configure webhook like `https://open.feishu.cn/open-apis/bot/v2/hook/{web_hook}` end `{web_hook}`
- `{web_hook}` must settings at `settings.feishu_webhook`or `PLUGIN_FEISHU_WEBHOOK`
- if set `Custom keywords` you can change `settings.feishu_msg_title` or `PLUGIN_FEISHU_MSG_TITLE`
- or set `Signature validation` by `settings.feishu_secret` or `PLUGIN_FEISHU_SECRET`

## Pipeline Settings (.drone.yml)

- `1.x` docker

sample config

```yaml
steps:
  - name: notification-feishu-group-robot
    image: sinlov/drone-feishu-group-robot:1.16.0
    pull: if-not-exists
    # image: sinlov/drone-feishu-group-robot:latest
    settings:
      # debug: true # plugin debug switch
      # ntp_target: "pool.ntp.org" # if not set will not sync ntp time
      feishu_webhook:
        # https://docs.drone.io/pipeline/environment/syntax/#from-secrets
        from_secret: feishu_group_bot_token
      feishu_secret:
        from_secret: feishu_group_secret_bot
      feishu_msg_title: "Drone CI Notification" # default [Drone CI Notification]
      # let notification card change more info see https://open.feishu.cn/document/ukTMukTMukTM/uAjNwUjLwYDM14CM2ATN
      feishu_enable_forward: true
    when:
      event: # https://docs.drone.io/pipeline/exec/syntax/conditions/#by-event
        - promote
        - rollback
        - push
        - pull_request
        - tag
      status: # only support failure/success,  both open will send anything
        - failure
        - success
```


full config

```yaml
steps:
  - name: notification-feishu-group-robot
    image: sinlov/drone-feishu-group-robot:latest
    settings:
      debug: true # plugin debug switch
      # ntp_target: "pool.ntp.org" # if not set will not sync ntp time
      timeout_second: 10 # default 10
      feishu_webhook:
        # https://docs.drone.io/pipeline/environment/syntax/#from-secrets
        from_secret: feishu_group_bot_token
      feishu_secret:
        from_secret: feishu_group_secret_bot
      # ignore last success branch by badges
      feishu_ignore_last_success_by_badges: true # will check branch badges, if success will not send message, tag build will not pass, default false
      feishu_ignore_last_success_branch: main # if not set, will use now drone build branch, and now branch status is started so not ignore, and if in tag mode, will not ignore
      # let notification card change more info see https://open.feishu.cn/document/ukTMukTMukTM/uAjNwUjLwYDM14CM2ATN
      feishu_msg_title: "Drone CI Notification" # default [Drone CI Notification]
      feishu_enable_forward: true
      drone_system_admin_token: # non-essential parameter 1.9.0+
        from_secret: drone_system_admin_token
      # ignore last success by distance
      feishu_ignore_last_success_by_admin_token_distance: 2 # if distance is 0 or now is tag will not ignore, use 1 will let notify build change to success, most use 2 to support tag release
      # oss info
      feishu_oss_host: "https://xxx.com" # OSS host for show oss info, if empty will not show oss info
      feishu_oss_info_send_result: ${DRONE_BUILD_STATUS} # append oss info must set success
      feishu_oss_info_user: "admin" # OSS user for show at card
      feishu_oss_info_path: "dist/foo/bar" # OSS path for show at card
      feishu_oss_resource_url: "https://xxx.com/s/xxx" # OSS resource url
      feishu_oss_page_url: "https://xxx.com/p/xxx" # OSS page url
      feishu_oss_page_passwd: "abc_xyz" # OSS password at page url, will hide PLUGIN_FEISHU_OSS_RESOURCE_URL when PAGE_PASSWD not empty 
    when:
      event: # https://docs.drone.io/pipeline/exec/syntax/conditions/#by-event
        - promote
        - rollback
        - push
        - pull_request
        - tag
      status: # only support failure/success,  both open will send anything
        - failure
        - success
```

- `1.x` drone-exec only support env
- download by [https://github.com/sinlov/drone-feishu-group-robot/releases](https://github.com/sinlov/drone-feishu-group-robot/releases) to get platform binary, then has local path
- binary path like `C:\Drone\drone-runner-exec\plugins\drone-feishu-group-robot.exe` can be drone run env like `EXEC_DRONE_FEISHU_GROUP_ROBOT_FULL_PATH`
- env:EXEC_DRONE_FEISHU_GROUP_ROBOT_FULL_PATH can set at file which define as [DRONE_RUNNER_ENVFILE](https://docs.drone.io/runner/exec/configuration/reference/drone-runner-envfile/) to support each platform to send feishu message

```yaml
steps:
  - name: notification-feishu-group-robot-exec # must has env EXEC_DRONE_FEISHU_GROUP_ROBOT_FULL_PATH and exec tools
    environment:
      PLUGIN_DEBUG: false
      # PLUGIN_NTP_TARGET: "pool.ntp.org" # if not set will not sync
      PLUGIN_TIMEOUT_SECOND: 10 # default 10
      PLUGIN_FEISHU_WEBHOOK:
        from_secret: feishu_group_bot_token
      PLUGIN_FEISHU_SECRET:
        from_secret: feishu_group_secret_bot
      PLUGIN_DRONE_SYSTEM_ADMIN_TOKEN: # non-essential parameter 1.9.0+
        from_secret: drone_system_admin_token
      # ignore last success by distance
      PLUGIN_FEISHU_IGNORE_LAST_SUCCESS_BY_ADMIN_TOKEN_DISTANCE: 2 # if distance is 0 or now is tag will not ignore, use 1 will let notify build change to success, most use 2 to support tag release
      # let notification card change more info see https://open.feishu.cn/document/ukTMukTMukTM/uAjNwUjLwYDM14CM2ATN
      PLUGIN_FEISHU_MSG_TITLE: "Drone CI Notification" # default [Drone CI Notification]
      PLUGIN_FEISHU_ENABLE_FORWARD: true
      # ignore last success branch by badges
      PLUGIN_FEISHU_IGNORE_LAST_SUCCESS_BY_BADGES: true # will check branch badges, if success will not send message, tag build will not pass, default false
      PLUGIN_FEISHU_IGNORE_LAST_SUCCESS_BRANCH: main # if not set, will use now drone build branch, and now branch status is started so not ignore, and if in tag mode, will not ignore
      # oss info to show
      PLUGIN_FEISHU_OSS_HOST: "https://xxx.com" # OSS host for show oss info, if empty will not show oss info
      PLUGIN_FEISHU_OSS_INFO_SEND_RESULT: ${DRONE_BUILD_STATUS} # append oss info must set success 
      PLUGIN_FEISHU_OSS_INFO_USER: "admin" # OSS user for show at card
      PLUGIN_FEISHU_OSS_INFO_PATH: "dist/foo/bar" # OSS path for show at card 
      PLUGIN_FEISHU_OSS_RESOURCE_URL: "https://xxx.com/s/xxx" # OSS resource url
      PLUGIN_FEISHU_OSS_PAGE_URL: "https://xxx.com/p/xxx" # OSS page url
      PLUGIN_FEISHU_OSS_PAGE_PASSWD: "abc_xyz" # OSS password at page url, will hide PLUGIN_FEISHU_OSS_RESOURCE_URL when PAGE_PASSWD not empty
    commands:
      - ${EXEC_DRONE_FEISHU_GROUP_ROBOT_FULL_PATH} `
        ""
    when:
      event: # https://docs.drone.io/pipeline/exec/syntax/conditions/#by-event
        - promote
        - rollback
        - push
        - pull_request
        - tag
      status: # only support failure/success,  both open will send anything
        - failure
        - success
```

### custom settings

- `settings.debug` or `PLUGIN_DEBUG` can open plugin debug mode
- `settings.timeout_second` or `PLUGIN_TIMEOUT_SECOND` can set send message timeout

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [convention-change-log](https://github.com/convention-change/convention-change-log) for commit guidelines.

## [1.16.0](https://github.com/sinlov/drone-feishu-group-robot/compare/1.15.0...v1.16.0) (2023-09-26)

### ✨ Features

* show build create and finish and interval time ([6d4305b6](https://github.com/sinlov/drone-feishu-group-robot/commit/6d4305b666b3c00bbf82b2ee6e1e92b280f26c08)), feat [#7](https://github.com/sinlov/drone-feishu-group-robot/issues/7)

### 👷‍ Build System

* github.com/sinlov/drone-info-tools v1.32.0 and github.com/antchfx/xmlquery v1.3.18 ([c91fcddb](https://github.com/sinlov/drone-feishu-group-robot/commit/c91fcddbdddde74b5926d31cb2698e91fc2ac88c))

## [1.15.0](https://github.com/sinlov/drone-feishu-group-robot/compare/1.14.1...v1.15.0) (2023-08-08)

### ✨ Features

* let feishu message card render HeadTemplateStyle support different event ([ef1a9791](https://github.com/sinlov/drone-feishu-group-robot/commit/ef1a97916f270230a3de438e87c09c7c600fe456))

* add feishu_card support DRONE_BUILD_EVENT tag pull_request ([8d49bd8b](https://github.com/sinlov/drone-feishu-group-robot/commit/8d49bd8b8d03875eeafbed3ad9a690848eb53d5b))

## [1.14.1](https://github.com/sinlov/drone-feishu-group-robot/compare/1.14.0...v1.14.1) (2023-08-07)

### 👷‍ Build System

* github.com/sinlov/drone-info-tools v1.31.0 ([e4de5e7f](https://github.com/sinlov/drone-feishu-group-robot/commit/e4de5e7f3dafdfa391d7e310a2e344c3bf2dd0a5))

## [1.14.0](https://github.com/sinlov/drone-feishu-group-robot/compare/1.13.0...v1.14.0) (2023-08-05)

### ✨ Features

* change format of show build link ([0139bf47](https://github.com/sinlov/drone-feishu-group-robot/commit/0139bf474ff4053a3684a2da8d760312f2e13e07))

### 👷‍ Build System

* change go version to 1.19.10 to build ([8259f6eb](https://github.com/sinlov/drone-feishu-group-robot/commit/8259f6ebb5ae37562ffabd719d88e42e045da4da))

## [1.13.0](https://github.com/sinlov/drone-feishu-group-robot/compare/1.12.0...v1.13.0) (2023-08-05)

### ✨ Features

* add title build Tag ([e73233c1](https://github.com/sinlov/drone-feishu-group-robot/commit/e73233c1778f162fb062e173f3935738e9c35c02))

* update feature doc ([20489e96](https://github.com/sinlov/drone-feishu-group-robot/commit/20489e96d8ce18fdbafcf213ebb4b787fa2c91bd))

### ♻ Refactor

* use new build system for fast build ([bac6a404](https://github.com/sinlov/drone-feishu-group-robot/commit/bac6a40437792b6ca06eed303e9cc06cf04300d1))

## [1.12.0](https://github.com/sinlov/drone-feishu-group-robot/compare/1.11.0...v1.12.0) (2023-08-04)

### ✨ Features

* support env DRONE_BUILD_DEBUG for debug of this tools as drone admin ([cf4df62d](https://github.com/sinlov/drone-feishu-group-robot/commit/cf4df62ddd2a9d66a4a4e4cff73781beb3f455bd))

### 👷‍ Build System

* change to go 1.18 and 1.18.10 to build ([f8c04d34](https://github.com/sinlov/drone-feishu-group-robot/commit/f8c04d341be92feea0bb0d869e7f8a81197e2642))

## [1.11.0](https://github.com/sinlov/drone-feishu-group-robot/compare/v1.10.0...v1.11.0) (2023-06-14)

### Features

* let feishu_ignore_last_success_by_admin_token_distance not pass tag to notifty ([ec1555b](https://github.com/sinlov/drone-feishu-group-robot/commit/ec1555bce120d9d24c910076ae1d38d83d45fc9b))

## [1.10.0](https://github.com/sinlov/drone-feishu-group-robot/compare/v1.9.0...v1.10.0) (2023-06-14)

### Features

* let checkIgnoreLastSuccessByAdminToken not pass when now drone build status not success ([1b27c6a](https://github.com/sinlov/drone-feishu-group-robot/commit/1b27c6a7f80ad2dc488360a627067fce82e9e3b0))

## [1.9.0](https://github.com/sinlov/drone-feishu-group-robot/compare/v1.8.0...v1.9.0) (2023-06-14)

### Features

* add drone_system_admin_token and feishu_ignore_last_success_by_admin_token_distance support ([0af41fc](https://github.com/sinlov/drone-feishu-group-robot/commit/0af41fcdf8dea1ae1cac61857ea736b2657fd52c))

* let checkIgnoreLastSuccessByAdminToken debug show more info ([5ce5047](https://github.com/sinlov/drone-feishu-group-robot/commit/5ce5047c0ed02010204fa7e75562495e8b42f9e9))

* let drone_system_admin_token or ENV: PLUGIN_DRONE_SYSTEM_ADMIN_TOKEN fail ([1d3de85](https://github.com/sinlov/drone-feishu-group-robot/commit/1d3de85a3795bde99a5b919526e3ef82f17160ab))

## [1.8.0](https://github.com/sinlov/drone-feishu-group-robot/compare/v1.7.0...v1.8.0) (2023-06-14)

### Features

* add feishu_ignore_last_success_by_badges ([a36c4ec](https://github.com/sinlov/drone-feishu-group-robot/commit/a36c4ec1f25eed6101a0965d9689777d44bfa7f0))

## [1.7.0](https://github.com/sinlov/drone-feishu-group-robot/compare/v1.6.0...v1.7.0) (2023-03-15)

### Features

* let detail show start time and finish time ([bc46a56](https://github.com/sinlov/drone-feishu-group-robot/commit/bc46a5691c8c627d5614e8d7819c2a710f4c679d))

## [1.6.0](https://github.com/sinlov/drone-feishu-group-robot/compare/v1.5.0...v1.6.0) (2023-02-11)

### Features

* fix Drone.Commit.Link compare not support, when tags Link get error ([c55854a](https://github.com/sinlov/drone-feishu-group-robot/commit/c55854acf74887699b754ff51ef8859937665087))

* update github.com/sinlov/drone-info-tools v1.7.0 and card render by Drone Tag ([6ee3f3c](https://github.com/sinlov/drone-feishu-group-robot/commit/6ee3f3cf01459e68aa97ece2358aed0973d2ff96))

## [1.6.0](https://github.com/sinlov/drone-feishu-group-robot/compare/v1.5.0...v1.6.0) (2023-02-11)


drone_info/droneInfo.go
package drone_info

const (
	DroneTimeFormatDefault = "2006-01-02-03-04-05"

	// EnvDroneCommitAuthorName
	// Provides the commit author name for the current running build. Note this is a user-defined value and may be empty or inaccurate.
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author-name/
	EnvDroneCommitAuthorName = "DRONE_COMMIT_AUTHOR_NAME"
	// EnvDroneCommitAuthorEmail
	// Provides the commit email address for the current running build. Note this is a user-defined value and may be empty or inaccurate.
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author-email/
	EnvDroneCommitAuthorEmail = "DRONE_COMMIT_AUTHOR_EMAIL"
	// EnvDroneCommitAuthor
	// Provides the commit author username for the current running build. This is the username from source control management system (e.g. GitHub username).
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author/
	EnvDroneCommitAuthor = "DRONE_COMMIT_AUTHOR"
	// EnvDroneCommitAuthorAvatar
	// Provides the commit author avatar for the current running build. This is the avatar from source control management system (e.g. GitHub).
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author-avatar/
	EnvDroneCommitAuthorAvatar = "DRONE_COMMIT_AUTHOR_AVATAR"
	EnvDroneCommitBranch       = "DRONE_COMMIT_BRANCH"
	EnvDroneCommitLink         = "DRONE_COMMIT_LINK"
	EnvDroneCommitMessage      = "DRONE_COMMIT_MESSAGE"
	EnvDroneCommitSha          = "DRONE_COMMIT_SHA"
	// EnvDroneCommitRef
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-ref/
	EnvDroneCommitRef = "DRONE_COMMIT_REF"
	// EnvDroneRepo
	// most is EnvDroneRepoNamespace / EnvDroneRepoName
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo/
	EnvDroneRepo     = "DRONE_REPO"
	EnvDroneRepoName = "DRONE_REPO_NAME"
	// EnvDroneRepoNamespace
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo-namespace/
	EnvDroneRepoNamespace = "DRONE_REPO_NAMESPACE"
	// EnvDroneGitHttpUrl
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-git-http-url/
	EnvDroneGitHttpUrl = "DRONE_GIT_HTTP_URL"
	// EnvDroneGitSshUrl
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-git-ssh-url/
	EnvDroneGitSshUrl = "DRONE_GIT_SSH_URL"
	EnvDroneRemoteUrl = "DRONE_REMOTE_URL"
	EnvDroneRepoOwner = "DRONE_REPO_OWNER"
	// EnvDroneRepoScm
	// must is: git hg
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo-scm/
	EnvDroneRepoScm = "DRONE_REPO_SCM"

	// EnvDroneBuildWorkSpace
	// drone’s working directory for a pipeline
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-workspace/
	EnvDroneBuildWorkSpace = "DRONE_WORKSPACE"
	EnvDroneBuildStatus    = "DRONE_BUILD_STATUS"
	EnvDroneBuildNumber    = "DRONE_BUILD_NUMBER"
	EnvDroneBuildLink      = "DRONE_BUILD_LINK"
	EnvDroneBuildEvent     = "DRONE_BUILD_EVENT"
	EnvDroneBuildStarted   = "DRONE_BUILD_STARTED"
	EnvDroneBuildFinished  = "DRONE_BUILD_FINISHED"

	EnvDroneTag = "DRONE_TAG"
	// EnvDroneTargetBranch
	// This environment variable can be used in conjunction with the source branch variable to get the pull request base and head branch.
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-target-branch/
	EnvDroneTargetBranch = "DRONE_TARGET_BRANCH"
	EnvDronePR           = "DRONE_PULL_REQUEST"
	EnvDroneDeployTo     = "DRONE_DEPLOY_TO"

	// EnvDroneStageStarted
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-started/
	EnvDroneStageStarted = "DRONE_STAGE_STARTED"
	// EnvDroneStageFinished
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-finished/
	EnvDroneStageFinished = "DRONE_STAGE_FINISHED"

	// EnvDroneStageMachine
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-machine/
	EnvDroneStageMachine = "DRONE_STAGE_MACHINE"
	// EnvDroneStageOs
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-os/
	EnvDroneStageOs = "DRONE_STAGE_OS"
	// EnvDroneStageArch
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-arch/
	EnvDroneStageArch = "DRONE_STAGE_ARCH"
	// EnvDroneStageVariant
	// Provides the target operating architecture variable for the current build stage. This variable is optional and is only available for arm architectures.
	// most is: "", or v7
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-variant/
	EnvDroneStageVariant = "DRONE_STAGE_VARIANT"
	// EnvDroneStageType
	// most use: docker
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-type/
	EnvDroneStageType = "DRONE_STAGE_TYPE"
	// EnvDroneStageKind
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-kind/
	EnvDroneStageKind = "DRONE_STAGE_KIND"
	// EnvDroneStageName
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-name/
	EnvDroneStageName = "DRONE_STAGE_NAME"

	// EnvDroneFailedStages
	// by env:DRONE_FAILED_STAGES
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-failed-stages/
	EnvDroneFailedStages = "DRONE_FAILED_STAGES"
	// EnvDroneFailedSteps
	// by env:DRONE_FAILED_STEPS
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-failed-steps/
	EnvDroneFailedSteps = "DRONE_FAILED_STEPS"

	// EnvDroneSystemVersion
	// Provides the version of the Drone server.
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-version/
	EnvDroneSystemVersion = "DRONE_SYSTEM_VERSION"
	// EnvDroneSystemHost
	// Provides the hostname used by the Drone server. This can be combined with the protocol to construct to the server url.
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-host/
	EnvDroneSystemHost = "DRONE_SYSTEM_HOST"
	// EnvDroneSystemHostName
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-hostname/
	EnvDroneSystemHostName = "DRONE_SYSTEM_HOSTNAME"
	// EnvDroneSystemProto
	// https or http
	// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-proto/
	EnvDroneSystemProto = "DRONE_SYSTEM_PROTO"
)

type (
	// Repo repo base info
	Repo struct {
		// ShortName
		// by env:DRONE_REPO_NAME
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo-name/
		ShortName string //  short name
		// GroupName
		// by env:DRONE_REPO_NAMESPACE
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo-namespace/
		GroupName string //  group name
		// FullName
		// by env:DRONE_REPO
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo/
		FullName string //  repository full name
		// OwnerName
		// by env:DRONE_REPO_OWNER
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo-owner/
		OwnerName string //  repo owner
		// Scm
		// by env:DRONE_REPO_SCM
		// must is: git hg
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-repo-scm/
		Scm string
		// RemoteURL
		// by env:DRONE_REMOTE_URL
		// Provides the git+https url that should be used to clone the repository
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-remote-url/
		RemoteURL string //  repo remote url
		// HttpUrl
		// by env:DRONE_GIT_HTTP_URL
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-git-http-url/
		HttpUrl string
		// SshUrl
		// by env:DRONE_GIT_SSH_URL
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-git-ssh-url/
		SshUrl string
		// Host
		// this from HttpUrl host
		Host string
		// HostName
		// this from HttpUrl hostname
		HostName string
	}

	// Build info
	Build struct {
		// WorkSpace
		// drone’s working directory for a pipeline
		// by env:DRONE_WORKSPACE
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-workspace/
		WorkSpace string
		Status    string //  providers the current build status
		Number    uint64 //  providers the current build number
		Tag       string //  providers the current build tag
		// TargetBranch
		// by env:DRONE_TARGET_BRANCH
		// This environment variable can be used in conjunction with the source branch variable to get the pull request base and head branch.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-target-branch/
		TargetBranch string
		// Link
		// by env:DRONE_BUILD_LINK
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-build-link/
		Link       string //  providers the current build link
		Event      string //  trigger event
		StartAt    uint64 //  build start at ( unix timestamp )
		FinishedAt uint64 //  build finish at ( unix timestamp )
		PR         string //  build pull request
		DeployTo   string //  build deploy to
		// FailedStages
		// by env:DRONE_FAILED_STAGES
		// Provides a comma-separate list of failed pipeline stages for the current running build.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-failed-stages/
		FailedStages string
		// FailedSteps
		// by env:DRONE_FAILED_STEPS
		// Provides a comma-separate list of failed pipeline steps.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-failed-steps/
		FailedSteps string
	}

	// Commit info
	Commit struct {
		// Link
		// by env:DRONE_COMMIT_LINK
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-link/
		Link    string //  providers the http link to the current commit in the remote source code management system(e.g.GitHub)
		Branch  string //  providers the branch for the current commit
		Message string //  providers the commit message for the current build
		// Sha
		// by env:DRONE_COMMIT_SHA
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-sha/
		Sha string //  providers the commit sha for the current build
		// Ref
		// by env:DRONE_COMMIT_REF
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-ref/
		Ref    string //  commit ref
		Author CommitAuthor
	}

	// Stage drone stage env
	Stage struct {
		// StartedAt
		// by env:DRONE_STAGE_STARTED
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-started/
		StartedAt uint64
		// FinishedAt
		// by env:DRONE_STAGE_FINISHED
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-finished/
		// FinishedTime
		// form StartedAt
		StartedTime string
		FinishedAt  uint64
		// FinishedTime
		// form FinishedAt
		FinishedTime string
		// Machine
		// by env:DRONE_STAGE_MACHINE
		// Provides the name of the host machine on which the pipeline is currently running.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-machine/
		Machine string
		// Os
		// by env:DRONE_STAGE_OS
		// Provides the target operating system for the current build stag
		// List of all possible values: darwin dragonfly freebsd linux netbsd openbsd solaris windows
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-os/
		Os string
		// Arch
		// by env:DRONE_STAGE_ARCH
		// Provides the platform architecture for the current build stage.
		// List of all possible values: 386 amd64 arm64 arm
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-arch/
		Arch string
		// Variant
		// by env:DRONE_STAGE_VARIANT
		// Provides the target operating architecture variable for the current build stage. This variable is optional and is only available for arm architectures.
		// most is: "", or v7
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-variant/
		Variant string
		// Type
		// by env:DRONE_STAGE_TYPE
		// This value is sourced from the type attribute in the yaml configuration file.
		// most of is: docker
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-type/
		Type string
		// Kind
		// by env:DRONE_STAGE_KIND
		// most of is: pipeline
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-kind/
		Kind string
		// Name
		// by env:DRONE_STAGE_NAME
		// most of is build
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-stage-name/
		Name string
	}

	// CommitAuthor commit author info
	CommitAuthor struct {
		// Username
		// by env:DRONE_COMMIT_AUTHOR_NAME
		// Provides the commit author name for the current running build. Note this is a user-defined value and may be empty or inaccurate.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author-name/
		Username string //  the author username for the current commit
		// Email
		// by env:DRONE_COMMIT_AUTHOR_EMAIL
		// Provides the commit email address for the current running build. Note this is a user-defined value and may be empty or inaccurate.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author-email/
		Email string //  providers the author email for the current commit
		// Name
		// by env:DRONE_COMMIT_AUTHOR
		// Provides the commit author username for the current running build. This is the username from source control management system (e.g. GitHub username).
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author/
		Name string //  providers the author name for the current commit
		// Avatar
		// by env:DRONE_COMMIT_AUTHOR_AVATAR
		// Provides the commit author avatar for the current running build. This is the avatar from source control management system (e.g. GitHub).
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-commit-author-avatar/
		Avatar string //  providers the author avatar for the current commit
	}

	DroneSystem struct {
		// Version
		// by env: DRONE_SYSTEM_VERSION
		// Provides the version of the Drone server.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-version/
		Version string
		// Host
		// by env:DRONE_SYSTEM_HOST
		// Provides the host used by the Drone server. This can be combined with the protocol to construct to the server url.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-host/
		Host string
		// HostName
		// by env:DRONE_SYSTEM_HOSTNAME
		// Provides the hostname used by the Drone server. This can be combined with the protocol to construct to the server url.
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-hostname/
		HostName string
		// Proto
		// by env:DRONE_SYSTEM_PROTO
		// Provides the protocol used by the Drone server. This can be combined with the hostname to construct to the server url.
		// https or http
		// @doc https://docs.drone.io/pipeline/environment/reference/drone-system-proto/
		Proto string
	}

	// Drone drone info
	Drone struct {
		Repo        Repo
		Build       Build
		Commit      Commit
		Stage       Stage
		DroneSystem DroneSystem
	}
)
